tests/pq/test_pgconn.py

```diff
@@ -389,6 +393,100 @@ def test_set_single_row_mode(pgconn):
     pgconn.set_single_row_mode()
 
 
+@contextlib.contextmanager
+def cancellable_query(pgconn: PGconn, monitor_conn: PGconn) -> Iterator[None]:
+    pgconn.send_query_params(b"SELECT pg_sleep($1)", [b"180"])
+    while True:
+        r = monitor_conn.exec_(
+            b"SELECT count(*) FROM pg_stat_activity"
+            b" WHERE query = 'SELECT pg_sleep($1)'"
+            b" AND state = 'active'"
+        )
+        assert r.status == pq.ExecStatus.TUPLES_OK
+        if r.get_value(0, 0) != b"0":
+            break
+
+        time.sleep(0.01)
+
+    yield None
+
+    res = pgconn.get_result()
+    assert res is not None
+    assert res.status == pq.ExecStatus.FATAL_ERROR
+    assert res.error_field(pq.DiagnosticField.SQLSTATE) == b"57014"
+    while pgconn.is_busy():
+        pgconn.consume_input()
+
+
+@pytest.mark.libpq(">= 17")
+def test_cancel_nonblocking(dsn):
+    # mimic test_cancel() from src/test/modules/libpq_pipeline/libpq_pipeline.c
+    def connect() -> PGconn:
+        conn = pq.PGconn.connect(dsn.encode())
+        if conn.status != pq.ConnStatus.OK:
+            pytest.fail(
+                f"bad connection: {conn.error_message.decode('utf8', 'replace')}"
+            )
+        return conn
+
+    conn, monitor_conn = connect(), connect()
+    conn.nonblocking = 1
+
+    # test PQcancel
+    with cancellable_query(conn, monitor_conn):
+        cancel = conn.get_cancel()
+        cancel.cancel()
+
+    # PGcancel object can be reused for the next query
+    with cancellable_query(conn, monitor_conn):
+        cancel.cancel()
+
+    del cancel
+
+    # test PQcancelBlocking
+    with cancellable_query(conn, monitor_conn):
+        cancel_conn = conn.cancel_conn()
+        assert cancel_conn.status == pq.ConnStatus.ALLOCATED
+        cancel_conn.blocking()
+        assert cancel_conn.status == pq.ConnStatus.OK  # type: ignore[comparison-overlap] # noqa: E501
+    cancel_conn.finish()
+    del cancel_conn
+
+    def wait(cancel_conn: PGcancelConn) -> None:
+        while True:
+            rv = cancel_conn.poll()
+            if rv == pq.PollingStatus.OK:
+                return
+            elif rv == pq.PollingStatus.READING:
+                select([cancel_conn.socket], [], [], 3)
+            elif rv == pq.PollingStatus.WRITING:
+                select([], [cancel_conn.socket], [], 3)
+            else:
+                pytest.fail(f"unexpected poll result: {rv}")
+        assert (
+            cancel_conn.status == pq.ConnStatus.OK
+        ), f"unexpected cancel connection status: {cancel_conn.error_message}"
+
+    # test PQcancelCreate and then polling with PQcancelPoll
+    with cancellable_query(conn, monitor_conn):
+        cancel_conn = conn.cancel_conn()
+        assert cancel_conn.status == pq.ConnStatus.ALLOCATED
+        cancel_conn.start()
+        assert cancel_conn.status == pq.ConnStatus.STARTED
+        wait(cancel_conn)
+        assert cancel_conn.status == pq.ConnStatus.OK
```

Comment:
Wouldn't be better to have several tests instead of a long one? cancellable_query might be a local fixture.

Reply:
Improved in 41e8b3a628805b4862c6d2bbace87c7cbcd72491.